File: Ahorn/entities/dreamspinner.jl
module IsaGrabBagDreamSpinner

using ..Ahorn, Maple

@mapdef Entity "isaBag/dreamSpinner" DreamSpinner(x::Integer, y::Integer, useOnce::Bool=false)

const placements = Ahorn.PlacementDict(
	"Dream Spinner (IsaGrabBag)" => Ahorn.EntityPlacement(
		DreamSpinner
	),
	"Dream Spinner (One Use) (IsaGrabBag)" => Ahorn.EntityPlacement(
		DreamSpinner,
		"rectangle",
		Dict{String, Any}(
			"useOnce" => true
		)
	)
)

function Ahorn.selection(entity::DreamSpinner)
	x, y = Ahorn.position(entity)
	return Ahorn.getSpriteRectangle("isafriend/danger/crystal/fg_dreamspinner.png", x, y)
end

function Ahorn.render(ctx::Ahorn.Cairo.CairoContext, entity::DreamSpinner, room::Maple.Room)
	Ahorn.drawSprite(ctx, "isafriend/danger/crystal/fg_dreamspinner.png", 0, 0)
end

end

File: Loenn/entities/dream_spinner.lua
local drawableSprite = require("structs.drawable_sprite")
local colors = require("consts.xna_colors")

local dreamSpinner = {}

dreamSpinner.name = "isaBag/dreamSpinner"
dreamSpinner.depth = -8500
dreamSpinner.placements = {
    {
        name = "default",
        data = {
            useOnce = false
        }
    },
    {
        name = "oneUse",
        data = {
            useOnce = true
        }
    }
}

local texture = "isafriend/danger/crystal/fg_dreamspinner"

function dreamSpinner.sprite(room, entity)
    local sprite = drawableSprite.fromTexture(texture, entity)
    
    if entity.useOnce then
        sprite:setColor(colors.Orange)
    end
    
    return sprite
end

function dreamSpinner.selection(room, entity)
    return utils.rectangle(entity.x - 10, entity.y - 10, 21, 21)
end

return dreamSpinner